### ISSUE TITLE
Allow assigning to rownames, colnames and dimnames for sparse matrices

### ISSUE THREAD

**Issue by Wainberg** [rpy2/rpy2#1087]
Currently you need to do `r.dimnames(rmatrix)[0] = ...` to set the rownames and `r.dimnames(rmatrix)[1] = ...`  to set the colnames. 

Would be great to support `rmatrix.rownames = ...`, `rmatrix.colnames = ...`, and `rmatrix.dimnames = ...` for consistency with dense matrices.

**Comment by lgautier** [comment]
The implementation of sparse matrices in R is only available through packages.  What R package do you have in mind?

I did play with an extension for rpy2 wrapping the R package Matrix (ttps://github.com/rpy2/rpy2-Matrix
), but between the complexity requires and the lack of users it stayed as an experiment and it might no longer work with recent version of `Matrix`.

**Comment by Wainberg** [comment]
Yeah it's the sparse matrices from the Matrix package (e.g. dgCMatrix). They're crucial for single-cell analysis, e.g. in the widely used Seurat package.

**Comment by lgautier** [comment]
Class definition in the R package `Matrix` use the "S4" OOP system (one of the two many ways to do OOP in R).

The base type in R is that this is an S4 object. This is mapped to the base type `rpy2.robjects.RS4` in rpy2. In the R package `Matrix` there is a base S4 class called `Matrix`, and methods such as `dimnames()`, `colnames()` and `rownames()` work through the notion of generic functions and a multiple dispatch system. 

Making the discovery of generics and all signatures associated under the multiple dispatch system fully automagic on the Python side, and map it through dynamic class and method generation on the Python side, is relatively complex. `rpy2` has something that can do it (see https://rpy2.github.io/doc/v3.5.x/html/robjects_oop.html#automated-r-in-python-class-definitions) if you want to try, but that not the default behavior. The Python package `rpy2-Matrix` (https://github.com/rpy2/rpy2-Matrix) is using that dynamic method generation.

However, here it does not seem that `colnames` and `rownames` are defined as generic by base R or `Matrix`. You could write child classes in for the rpy2 mapped classes and implement the methods you are missing.
